client/api/README.md

```diff
@@ -0,0 +1,253 @@
+# CSI-proxy's API
+
+## Overview
+
+CSI-proxy's API is a GRPC, versioned API.
+
+The server exposes a number of API groups, all independent of each other. Additionally, each API group has one or more versions. Each version in each group listens for GRPC messages on a Windows named pipe of the form `\\.\\pipe\\csi-proxy-<api_group_name>-<version>` (e.g. `\\.\\pipe\\csi-proxy-filesystem-v2alpha1` or `\\.\\pipe\\csi-proxy-iscsi-v1`).
+
+APIs are defined by protobuf files; each API group should live in its own directory under `client/api/<api_group_name>` in this repo's root (e.g. `client/api/iscsi`), and then define each of its version in `client/api/<api_group_name>/<version>/api.proto` files (e.g. `client/api/iscsi/v1/api.proto`). Each `proto` file should define exactly one RPC service.
+
+Internally, there is only one server `struct` per API group, that handles all the versions for that API group. That server is defined in this repo's `internal/server/<api_group_name>` (e.g. `internal/server/iscsi`) go package. This go package should follow the following pattern:
+
+<a name="serverPkgTree"></a>
+```
+internal/server/<api_group_name>
+├── internal
+│   └── types.go
+└── server.go
+```
+where `types.go` should contain the internal types corresponding to the various protobuf types for that API group - these internal structures must be able to represent all the different versions of the API. For example, given a `dummy` API group with two versions defined by the following `proto` files:
+
+`client/api/dummy/v1alpha1/api.proto`
+```proto
+syntax = "proto3";
+
+package v1alpha1;
+
+service Dummy {
+    // ComputeDouble computes the double of the input. Real smart stuff!
+    rpc ComputeDouble(ComputeDoubleRequest) returns (ComputeDoubleResponse) {}
+}
+
+message ComputeDoubleRequest{
+    int32 input32 = 1;
+}
+
+message ComputeDoubleResponse{
+    int32 response32 = 1;
+}
+```
+
+and
+
+`client/api/dummy/v1/api.proto`
+```proto
+syntax = "proto3";
+
+package v1;
+
+service Dummy {
+    // ComputeDouble computes the double of the input. Real smart stuff!
+    rpc ComputeDouble(ComputeDoubleRequest) returns (ComputeDoubleResponse) {}
+}
+
+message ComputeDoubleRequest{
+    // we changed in favor of an int64 field here
+    int64 input = 2;
+}
+
+message ComputeDoubleResponse{
+    int64 response = 2;
+    
+    // set to true if the result overflowed
+    bool overflow = 3;
+}
+```
+
+then `internal/server/dummy/internal/types.go` could look something like:
+```go
+type ComputeDoubleRequest struct {
+	Input int64
+}
+
+type ComputeDoubleResponse struct {
+	Response int64
+	Overflow bool
+}
+```
+and then the API group's server (`internal/server/dummy/server.go`) needs to define the callbacks to handle requests for all API versions, e.g.:
+```go
+type Server struct{}
+
+func (s *Server) ComputeDouble(ctx context.Context, request *internal.ComputeDoubleRequest, version apiversion.Version) (*internal.ComputeDoubleResponse, error) {
```

Comment:
The parameter of "version" is not used in the function?

Reply:
Yes. That's required to fulfill the auto-generated interface that the generated versioned servers expect (https://github.com/kubernetes-csi/csi-proxy/pull/7/files#diff-fed01b95f0f43874e57816bfe316b1f7R18).

The idea here is that API devs can use that to make adjustments based on the API version the request was made against. The preferred way to handle that would be through custom conversion functions though.